### ISSUE TITLE
Regent: index launch optimizer failure

### ISSUE THREAD

**Issue by syamajala** [StanfordLegion/legion#767]
The following fails to compile:

```
import "regent"
local c = regentlib.c

fspace AFields
{
  Field1 : double,
  Field2 : double,
  Field3 : double,
  Field4 : double
}

struct rAFields
{
  A : double,
  B : double,
}

task test(rA : region(ispace(int3d), rAFields),
          rB : region(ispace(int3d), double))
where
  writes(rA.{A}),
  reads(rA.{B}),
  reads(rB)
do
end

task main()

  var grid = ispace(int3d, {8, 8, 8})
  var rA = region(grid, AFields)
  var rB = region(grid, double)

  var colors = ispace(int3d, {2, 2, 2})
  var lp_rA = partition(equal, rA, colors)
  var lp_rB = partition(equal, rB, colors)

  for color in colors do
    test(lp_rA[color].{A=Field1,
                       B=Field2},
         lp_rB[color])
  end
end

regentlib.start(main)
```

It results in this error:
```
...dev/legion/language/src/regent/optimize_index_launches.t:237: attempt to index local 'other_region_type' (a boolean value)
stack traceback:
        ...dev/legion/language/src/regent/optimize_index_launches.t:237: in function 'analyze_noninterference_previous'
        ...dev/legion/language/src/regent/optimize_index_launches.t:1121: in function 'optimize_loop_body'
        ...dev/legion/language/src/regent/optimize_index_launches.t:1230: in function 'map_node_postorder'
        /home/seshu/dev/legion/language/src/common/ast.t:520: in function 'map_node_postorder'
        /home/seshu/dev/legion/language/src/common/ast.t:511: in function 'map_node_postorder'
        /home/seshu/dev/legion/language/src/common/ast.t:511: in function </home/seshu/dev/legion/language/src/common/ast.t:505>
        /home/seshu/dev/legion/language/src/regent/passes_hooks.t:41: in function 'optimize'
        /home/seshu/dev/legion/language/src/regent/passes.t:52: in function </home/seshu/dev/legion/language/src/regent/passes.t:48>
        index_launch.rg:27: in main chunk
```
The only way to get it to compile is to use __forbid(__index_launch).

**Comment by elliottslaughter** [comment]
Looks like an issue in field polymorphism? CC @magnatelee

**Comment by syamajala** [comment]
I realized I need to be able to index space launch tasks in order to use the dumping code on multiple nodes otherwise task->index_point doesn't get filled out and I don't know what subregion we're dumping values for in S3D.

**Comment by lightsighter** [comment]
I'm not sure if Regent supports this, but Legion will let you set an arbitrary point on a single task launch so you can give it some context from the calling task:

https://gitlab.com/StanfordLegion/legion/-/blob/master/runtime/legion.h#L1464

**Comment by magnatelee** [comment]
This should be fixed by 47245b1a. @syamajala Please pull and try again.

**Comment by syamajala** [comment]
This seems to work for the posted example.
